app/labor/markdown_fixer.rb

```diff
@@ -47,11 +47,29 @@ def split_tags(markdown)
     end
 
     def underscores_in_usernames(markdown)
-      markdown.gsub(/(@)(_\w+)(_)/, '\1\\\\\2\\\\\3')
+      return markdown unless markdown.include?("@")
+
+      traverser = MarkdownTraverser.new(markdown)
+      traverser.each do |line|
+        unless traverser.in_codeblock?
+          escape_underscored_username_in_line!(line)
+        end
+      end.join
     end
 
     private
 
+    # Match @_username_ that is not preceded by backtick
+    USERNAME_WITH_UNDERSCORE_REGEXP = /(?<!`)@_\w+_/.freeze
+
+    # Escapes underscored username that is not in code
+    def escape_underscored_username_in_line!(line)
+      line.scan(USERNAME_WITH_UNDERSCORE_REGEXP).each do |to_escape|
```

Comment:
Instead of `scan` + `each` and then explicitly returning `line`, could we use the block form of `gsub` here? Something like (untested)

```ruby
line.gsub(USERNAME_WITH_UNDERSCORE_REGEXP) { |to_escape| to_escape.gsub("_", "\\_") }
```

Reply:
I would like this too, but the `line` could be

```text
@_dev_ `@_dev_`
```

hence we cannot replace all underscores in a line.

Reply:
This is not replacing all underscores in a line, this is replacing all underscores in  every `to_escape`, which is everything that matches `USERNAME_WITH_UNDERSCORE_REGEXP`:

```ruby
line = '@_dev_ `@_dev_`'
line.gsub(USERNAME_WITH_UNDERSCORE_REGEXP) { |to_escape| to_escape.gsub("_", "\\_") }
#=> "@\\_dev\\_ `@_dev_`"
```

Isn't this what you want, i.e. escape the username outside backticks but leave the one inside unchanged?

Reply:
Sorry I overlooked from my last comment. But I did give this a try and other places need to modify as well but I don‘t have time to look into why this does not work. What‘s your concern on the current implementation? I‘m happy to receive a PR against this feature branch.

Reply:
No concern, it just seemed like it's using a more complicated approach than necessary. I'll trust you if you say it doesn't work, so let's just leave it as is.